src/hooks/useJournalData.ts

```diff
@@ -0,0 +1,252 @@
+        LocalStorageKeys.GATEWAY,
+        getStoredGatewayAuthConfig()
+    );
+
+    const journalClient = useMemo(() => {
+        if (gatewayConfig?.gateway_url && gatewayConfig.token) {
+            const authToken = gatewayConfig.token;
+            const baseUrl = new URL(gatewayConfig.gateway_url);
+
+            return new JournalClient(baseUrl, authToken);
+        } else {
+            return null;
+        }
+    }, [gatewayConfig]);
+
+    const fetcher = useCallback(
+        (_url: string) => {
+            if (journalClient && collectionName) {
+                const journalSelector = new JournalSelector().collection(
+                    collectionName
+                );
+                return journalClient.list(journalSelector).then((result) => {
+                    const journals = result.unwrap();
+
+                    return {
+                        journals: journals.length > 0 ? journals : [],
+                    };
+                });
+            } else {
+                return null;
+            }
+        },
+        [collectionName, journalClient]
+    );
+
+    return useSWR(
+        collectionName
+            ? `journals-${
+                  gatewayConfig?.gateway_url ?? '__missing_gateway_url__'
+              }`
+            : null,
+        fetcher,
+        {
+            // TODO (data preview refresh) no polling right now we should add a manual refresh button
+            errorRetryInterval: undefined,
+            refreshInterval: undefined,
+            revalidateOnFocus: false,
+            onError: (error) => {
+                if (
+                    session &&
+                    (`${error}`.includes(ErrorFlags.TOKEN_INVALID) ||
+                        `${error}`.includes(ErrorFlags.TOKEN_NOT_FOUND))
+                ) {
+                    const prefixes: string[] = grantDetails.map(
+                        ({ object_role }) => object_role
+                    );
+
+                    getGatewayAuthConfig(prefixes, session.access_token)
+                        .then(([response]) => {
+                            setGatewayConfig(response);
+                        })
+                        .catch((configError) => Promise.reject(configError));
+                }
+
+                console.error(error);
+
+                return Promise.reject(error);
+            },
+        }
+    );
+};
+
+async function* streamAsyncIterator<T>(stream: ReadableStream<T>) {
+    // Get a lock on the stream
+    const reader = stream.getReader();
+
+    try {
+        // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+        while (true) {
+            // Read from the stream
+            const { done, value } = await reader.read();
+            // Exit if we're done
+            if (done) return;
+            // Else yield the chunk
+            yield value;
+        }
+    } finally {
+        reader.releaseLock();
+    }
+}
+
+const INCREMENT = 1024 * 10;
```

Comment:
nit: this name is really vague, and mixing of the power of 2 and 10 makes it seem like the specific value is really important. Maybe add a comment?

Reply:
Super not important actually, this is just a number of how many bytes to read back if the amount of data returned isn't enough to satisfy the desired number of rows

Reply:
comment was added.